### PR TITLE
Implement sample/gather for D3D12 shader debugging.

### PR DESCRIPTION
**Pull request by skarolewics**
This creates a VS/PS pair to sample the requested texture. It uses the slots that are already set, so the existing root signature can be used.

### CODE REVIEW

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp [hunk 1]
```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
...
+  if(FAILED(hr))
+  {
+    RDCERR("Failed to create texture for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(vsBlob);
+    SAFE_RELEASE(psBlob);
+    return false;
+  }
+
+  // TODO: This work to create a staging buffer, transition  the render target, and call
+  // CopyTextureRegion work should go into a function in the debug manager, similar to GetBufferData
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: Could you use `D3D12Replay::GetTextureData` for this?

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp [hunk 2]
```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
...
+  {
+    RDCERR("Failed to create PSO for shader debugging HRESULT: %s", ToStr(hr).c_str());
+    SAFE_RELEASE(vsBlob);
+    SAFE_RELEASE(psBlob);
+    ShaderVariable lookupResult("tex", 0.0f, 0.0f, 0.0f, 0.0f);
+    output = lookupResult;
+    return true;
+  }
+
+  ID3D12GraphicsCommandListX *cmdList = m_pDevice->GetDebugManager()->ResetDebugList();
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_shaderdebug.cpp:614]: You can release the blobs here right after creating the pipeline state, they're consumed inside that call and aren't used afterwards. Might make the error handling a little easier in the rest of the function not having to hold onto them.

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp [hunk 3]
```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
...
+  pipeDesc.SampleDesc.Count = 1;
+  pipeDesc.SampleDesc.Quality = 0;
+
+  pipeDesc.InputLayout.NumElements = 0;
+  pipeDesc.InputLayout.pInputElementDescs = NULL;
+  pipeDesc.IBStripCutValue = D3D12_INDEX_BUFFER_STRIP_CUT_VALUE_DISABLED;
+  pipeDesc.PrimitiveTopologyType = D3D12_PRIMITIVE_TOPOLOGY_TYPE_TRIANGLE;
+  pipeDesc.RTVFormats.NumRenderTargets = 1;
+  pipeDesc.RTVFormats.RTFormats[0] = retFmt;
+  pipeDesc.DSVFormat = DXGI_FORMAT_UNKNOWN;
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: I think there's a few more states to patch - we'd need to disable blending, ensure the SampleMask is correct, sanitise depth and stencil testing, the new view instancing too might break.

Thinking about it, is it easier to only pull in what we actually want? do we need anything from the original PSO apart from the root signature?

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp [hunk 4]
```diff
@@ -131,8 +131,645 @@ bool D3D12DebugAPIWrapper::CalculateSampleGather(
...
+
+  std::string sampleProgram;
+
+  UINT texSlot = resourceData.slot;
+  UINT sampSlot = samplerData.slot;
+
+  if(opcode == OPCODE_SAMPLE || opcode == OPCODE_SAMPLE_B || opcode == OPCODE_SAMPLE_D)
+  {
+    sampleProgram = textureDecl + " : register(t" + std::to_string(texSlot) + ");\n" + samplerDecl +
+                    " : register(s" + std::to_string(sampSlot) + ");\n\n";
```
Comment (baldurk) [inline renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: Prefer not using `std::to_string`, if you need an int-to-string then `ToStr` will do the trick, or for more complex formatting like this you might want to use `StringFormat::Fmt`.

(I know the D3D11 and this code uses `StringFormat::snprintf` which is fine for now, but that's super old and predates the existance of `StringFormat::Fmt`. I plan on removing the `snprintf` variant so using `Fmt` is the way to go for formatting in purely new code).

**Review by baldurk — CHANGES_REQUESTED** [review]
Looks good! I've left a few notes but nothing critical just cleanups/potential improvements.

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp (hunk 3 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: Yeah, just the root signature. Switched to just creating the PSO from scratch.

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp (hunk 4 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: Good to know. Migrated to StringFormat::Fmt for the places that were using std::to_string, and removed a few snprintf calls as well.

File: renderdoc/driver/d3d12/d3d12_shaderdebug.cpp (hunk 1 above)
Comment (skarolewics) [reply renderdoc/driver/d3d12/d3d12_shaderdebug.cpp]: This is exactly what I was looking for. Thanks!

**Review by baldurk — APPROVED** [review]

**Pull request merged**